src/mm-input-mask/mm-input-mask.js

```diff
@@ -0,0 +1,396 @@
+			this.parseMask();
+
+		},
+
+		domReady: function() {
+			this.updateGroups();
+		},
+
+		attached: function() {
+			this.$.input.addEventListener("focus", this.handleFocus.bind(this));
+			this.$.input.addEventListener("blur", this.handleBlur.bind(this));
+		},
+
+		detached: function() {
+			this.$.input.removeEventListener("focus");
+			this.$.input.removeEventListener("blur");
+		},
+
+		parseMask: function() {
+			this.sepsLen = 0; //used to check 'empty' state for placeholder styling
+			var nodes = DataUtils.clone( this.$.content.getDistributedNodes() ),
+				groups = [],
+				seps = [];
+			this.maskConfig = nodes.map(function(node, index) {
+				switch(node.nodeName) {
+					case "GROUP":
+						var rng = _groupRange(node.getAttribute("size")),
+							rule = this.rules[node.getAttribute("rule")],
+							restrict = this.restrict[node.getAttribute("restrict")],
+							auto = node.getAttribute("autofill"),
+							args = node.textContent.trim(),
+							style = {
+								width:"auto"
+							};
+						if (restrict !== _restrict.numeric) style = {};
+
+						var o = {
+							index:index,
+							max:rng[1],
+							min:rng[0], 
+							rule:rule, 
+							restrict:restrict, 
+							args:args, 
+							auto:auto,
+							type:_types.GROUP, 
+							style:style,
+							loaded: false,
+							contentWidth:this.contentWidthHandler
+						};
+						groups.push(o);
+						return o;
+
+					case "SEP":
+						var chars = node.getAttribute("chars");
+						seps.push(chars);
+						this.sepsLen += chars.length;
+						return {
+							type:_types.SEP, 
+							value:chars, 
+							style:{
+								sep:true, 
+								placeholder:true
+							}
+						};
+				}
+			}, this);
+			this.seps = seps;
+			this.groups = groups;
+
+		},
+
+		contentWidthHandler: function(style, val, placeholder, loaded) {
+			var w = 0.0;
+			var check = val || placeholder || "";
+			while(check.length < this.max) {
+				check += "S";
+			}
+			if (!val && placeholder) {
+				w = Measure.textWidth(null, check, "italic 13px Arimo");
+			} else {
+				w = Measure.textWidth(null, check, "13px Arimo");
+			}
+			style.width = w + "px";
+			return style;
+		},
+
+		placeholderChanged: function(oldPlaceholder, newPlaceholder) {
+			this.chunkValue(this.placeholder, "placeholder");
+		},
+
+		valueChanged: function(oldValue, newValue) {
+			if (newValue) {
+				var placeholder = newValue.length <= this.sepsLen;
+				this.maskConfig.filter(function(m) { return m.type === _types.SEP}).forEach(function(sep) {
+					sep.style = {sep:true, placeholder:placeholder};
+				},this);
+			}
+			if (this.ignoreInternal) {
+				this.ignoreInternal = false;
+				return;
+			}
+			this.chunkValue(newValue);
+		},
+
+		chunkValue: function(value, type) {
+			if (!type) type = "value";
+			//clear case
+			if (value === "" || value === null) {
+				for(var i=0; i<this.groups.length; i++) {
+					var g = this.groups[i];
+					this.maskConfig[g.index][type] = "";
```

Comment:
Is this.maskConfig[g.index] === g ever false?  Why not just use g/this.groups[i]?